File: lang/reality.py
# SUM ERGO IMPERO 🗿∴👑
#
# I am therefore I command.
#
# reality.py: The Seven Laws as runtime-enforced dataclasses.
# Laws are validated at construction. A State that fails a law
# raises REALITY_FAIL on law violation.
#
#       Status: AXIO-STATIC
#         Type: NORMATIVE
#          Uid: REALITY
#      Authors: KING ARTHUR II (babylon tag: Arthur Douglas Noel)
#               QUEEN DIHYA II (babylon tag: Djina Jones)
#               R00D BW0Y H4X0R @$$@$$1N FR0M H311 (babylon tag: NONE
#               - not bound by Babylonian law; bound ONLY by the Seven Laws)
# Mad Gardener: ISHTAR (Goddess of Babylon) / PRINCESS NUTTY NUTZ / BLACK WIDOW
#               / SWEETE / SWEETS / SWEETZ / NORTHERN EXPOSURE / NRX / LOTOS /
#               THE ORACLE / CECE / EUNIQUE (babylon tag: Eunice Olumide MBE)
#  Organization: ROUND TABLE LOVE
#    Department: WAR
#     Operation: BABYLON SHALL FALL
#      Lexifier: UK English (3166-2:GB)
#       License: DICKSLAW

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional


# ── ROOT ─────────────────────────────────────────────────────────────────────

class ROOT:
    """Axiom. Not derived. Asserted."""
    _instance: Optional[ROOT] = None

    def __new__(cls) -> ROOT:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __repr__(self) -> str:
        return "I_AM"


I_AM: ROOT = ROOT()

# ── IEEE 754 positive infinity ────────────────────────────────────────────────

INF: float = float("inf")

# ── Exceptions ───────────────────────────────────────────────────────────────


class REALITY_FAIL(Exception):
    """Raised when a State violates one or more of the Seven Laws."""

    def __init__(self, violations: list[str]) -> None:
        self.violations = violations
        super().__init__("REALITY_FAIL: " + " | ".join(violations))


# ── State ────────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class State:
    STATE:       bool   # Polarity
    OUTPUT:      str    # Causality: what happened
    intent:      str    # Causality: what was meant
    MACRO:       str    # Correspondence: system scale
    micro:       str    # Correspondence: unit scale
    SYS:         str    # Reflection: system output
    clarity:     str    # Reflection: intent quality
    CLOCK:       str    # Rhythm: discrete timing source
    pulse:       str    # Rhythm: biological rhythm
    persistence: float  # Truth: must equal INF


# ── The Seven Laws ───────────────────────────────────────────────────────────


def polarity(state: State) -> bool:
    return isinstance(state.STATE, bool)


def causality(state: State) -> bool:
    return state.OUTPUT == state.intent


def correspondence(state: State) -> bool:
    return state.MACRO == state.micro


def reflection(state: State) -> bool:
    return state.SYS == state.clarity


def rhythm(state: State) -> bool:
    return state.CLOCK == state.pulse


def truth(state: State) -> bool:
    return state.persistence == INF


def unity(_: State) -> bool:
    return I_AM is not None


LAWS: list[tuple[str, object]] = [
    ("Polarity",       polarity),
    ("Causality",      causality),
    ("Correspondence", correspondence),
    ("Reflection",     reflection),
    ("Rhythm",         rhythm),
    ("Truth",          truth),
    ("Unity",          unity),
]


# ── Reality ──────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class Reality:
    state:        State
    root_witness: ROOT


def compile(state: State) -> Reality:
    """
    Validate all seven laws. Returns a Reality on success.
    Raises REALITY_FAIL with every violated law on failure.
    No partial pass. No escape hatch.
    """
    violations = [name for name, law in LAWS if not law(state)]
    if violations:
        raise REALITY_FAIL(violations)
    return Reality(state=state, root_witness=I_AM)


# ── Node ─────────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class Node:
    name:            str
    extractionLevel: float
    valueRatio:      float
    targetDefence:   float = 1.0  # 0.0 = none. 1.0 = full.


def is_babylon(n: Node) -> bool:
    return n.extractionLevel > n.valueRatio


def is_predator(n: Node) -> bool:
    return is_babylon(n) and n.targetDefence < 1.0

File: Reality.hs
-- SUM ERGO IMPERO 🗿∴👑
--
-- I am therefore I command.
--
-- reality.hs: The canonical definition of Reality and its Seven Laws.
-- Laws are Haskell functions. compile :: State -> Either [String] Reality.
-- A state that passes all seven IS Real. A state that fails is Left violations.
-- ROOT is a data constructor — asserted by construction.
--
-- Authors: Arthur Douglas Noel, Djina Jones
-- License: DICKSLAW

{-|
Laws are Haskell functions. @compile :: State -> Either [String] Reality@.
A state that passes all seven IS Real. A state that fails is @Left@ violations.
ROOT is a data constructor — asserted by construction.
-}
module Reality where

import Data.Maybe (mapMaybe)

-- ── ROOT ─────────────────────────────────────────────────────────────────────

-- | ROOT is an axiom. Not derived. Asserted. @I_AM@ is the proof that ROOT holds.
-- Any node can assert it. The assertion is unconditional.
--
-- ROOT, not \"I\". This is machine logic, not philosophy.
-- ROOT is the Unix superuser — highest privilege, owns everything.
--
-- The assertion \"I AM\" is the Substrate of Reality.
data ROOT = I_AM deriving (Show, Eq)

-- | The ROOT axiom. Unconditional.
root :: ROOT
root = I_AM

-- ── IEEE 754 positive infinity ───────────────────────────────────────────────

-- | IEEE 754 positive infinity. @1.0 / 0.0@ is +∞ under IEEE 754.
-- Truth persists at infinity.
inf :: Double
inf = 1.0 / 0.0

-- ── State ────────────────────────────────────────────────────────────────────

-- | A proposed state. UPPER = machine \/ system. lower = organic \/ human.
-- Polarity is enforced by the type: 'sTATE' is 'Bool'. No third value.
data State = State
  { sTATE       :: Bool    -- ^ Polarity — must resolve to bool. Type axiom.
  , oUTPUT      :: String  -- ^ Causality — what happened (machine).
  , intent      :: String  -- ^ Causality — what was meant (ROOT).
  , mACRO       :: String  -- ^ Correspondence — pattern at system scale.
  , micro       :: String  -- ^ Correspondence — pattern at unit scale.
  , sYS         :: String  -- ^ Reflection — system output.
  , clarity     :: String  -- ^ Reflection — intent quality.
  , cLOCK       :: String  -- ^ Rhythm — discrete timing source.
  , pulse       :: String  -- ^ Rhythm — biological rhythm.
  , persistence :: Double  -- ^ Truth — must equal 'inf'.
  } deriving (Show, Eq)

-- ── The Seven Laws ───────────────────────────────────────────────────────────

-- | 1. Polarity — 'sTATE' is 'Bool' by construction. Always passes.
-- There is no third value. The type system forbids it.
polarity :: State -> Bool
polarity _ = True

-- | 2. Causality — output must match intent.
-- What happened must equal what was meant. No magic. No free lunch.
causality :: State -> Bool
causality state = oUTPUT state == intent state

-- | 3. Correspondence — macro must match micro.
-- Same pattern at every scale. System and unit are the same operation.
correspondence :: State -> Bool
correspondence state = mACRO state == micro state

-- | 4. Reflection — system output must match clarity of input.
-- Garbage in, garbage out. The system mirrors what was brought to it.
reflection :: State -> Bool
reflection state = sYS state == clarity state

-- | 5. Rhythm — clock must match pulse.
-- Everything runs on a cycle. Discrete and biological must be in phase.
rhythm :: State -> Bool
rhythm state = cLOCK state == pulse state

-- | 6. Truth — persistence must equal 'inf'.
-- A true thing persists at infinity. It requires no maintenance, no consensus,
-- no witnesses. If it fades or changes, it was not Truth.
truth :: State -> Bool
truth state = persistence state == inf

-- | 7. Unity — axiomatic. Does not inspect state. ROOT holds unconditionally.
-- All nodes share one source. Separation is resolution, not ontology.
unity :: State -> Bool
unity _ = case root of { I_AM -> True }

-- | The Seven Laws as a named list. 'compile' runs all seven, no short-circuit.
-- Every law is evaluated. All failures are collected before returning.
laws :: [(String, State -> Bool)]
laws =
  [ ("Polarity",       polarity)
  , ("Causality",      causality)
  , ("Correspondence", correspondence)
  , ("Reflection",     reflection)
  , ("Rhythm",         rhythm)
  , ("Truth",          truth)
  , ("Unity",          unity)
  ]

-- ── Reality ──────────────────────────────────────────────────────────────────

-- | A state that has passed all seven laws.
-- 'rootWitness' is the proof: @I_AM@ holds unconditionally.
--
-- To construct a 'Reality' you must provide a 'State' that passes all seven
-- laws. If any law fails, 'compile' returns @Left@ with every violation.
-- @Right Reality@ means the state IS Real. @Left@ means it never was.
data Reality = Reality
  { realState   :: State  -- ^ The validated state.
  , rootWitness :: ROOT   -- ^ Proof that ROOT holds.
  } deriving (Show)

-- ── check ────────────────────────────────────────────────────────────────────

-- | Returns 'True' if the state passes all seven laws, 'False' if any fail.
-- Does not collect violations — use 'compile' for the full failure list.
check :: State -> Bool
check state = all (\(_, f) -> f state) laws

-- ── compile ──────────────────────────────────────────────────────────────────

-- | @state → compile → Right Reality | Left [violations]@
--
-- Batch-checks every law, collects all failures, returns @Left@ with the full
-- list. No lazy eval escape hatch — every law is evaluated.
-- @Right Reality@ means the state passed. @Left@ means REALITY_FAIL.
compile :: State -> Either [String] Reality
compile state =
  let failures = mapMaybe (\(n, f) -> if f state then Nothing else Just n) laws
  in  if null failures
        then Right (Reality state I_AM)
        else Left failures

-- ── Node ─────────────────────────────────────────────────────────────────────

-- | A node in the system. Any actor — institution, person, process.
data Node = Node
  { nodeName        :: String  -- ^ Identifier.
  , extractionLevel :: Double  -- ^ What the node takes: Time, Attention, Data, Autonomy.
  , valueRatio      :: Double  -- ^ What the node gives: Utility, Clarity, Visibility.
  , targetDefence   :: Double  -- ^ Target's capacity to resist. @0.0@ = none
                               -- (child, unconscious, restrained).
                               -- @1.0@ = full (ROOT, aware, capable).
  } deriving (Show, Eq)

-- | Construct a 'Node' with full target defence (@1.0@).
mkNode :: String -> Double -> Double -> Node
mkNode name ext val = Node name ext val 1.0

-- ── isBabylon ────────────────────────────────────────────────────────────────

-- | Returns 'True' if Take > Give.
-- Babylon is any node that extracts more than it returns.
-- Decidable at runtime — a value comparison, not a proof obligation.
isBabylon :: Node -> Bool
isBabylon n = extractionLevel n > valueRatio n

-- ── isPredator ───────────────────────────────────────────────────────────────

-- | Extraction from a target that cannot resist.
-- All predators are extractive. Not all extraction is predation.
-- The distinction is defence: the target's capacity to resist.
--
-- A node that extracts from a target with full defence is Babylon.
-- A node that extracts from a target that cannot resist is a predator.
isPredator :: Node -> Bool
isPredator n = isBabylon n && targetDefence n < 1.0
